### PR TITLE
Serve static files from Azure CDN

### PR DESCRIPTION
**Pull request by ganniaiuei**
Following: [issue#112](https://github.com/aspnet/live.asp.net/issues/112)

### CODE REVIEW

File: src/live.asp.net/AppSettings.cs [hunk 1]
```diff
@@ -16,5 +16,7 @@ public class AppSettings
         public string AzureStorageBlobName { get; set; }
 
         public string AzureStorageContainerName { get; set; }
+
+		public string CDNUrl { get; set; }
```
Comment (hishamco) [inline src/live.asp.net/AppSettings.cs]: You miss one space before

File: src/live.asp.net/Startup.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@
 using Microsoft.IdentityModel.Protocols.OpenIdConnect;
 using live.asp.net.Formatters;
 using live.asp.net.Services;
+using Microsoft.Net.Http.Headers;
```
Comment (hishamco) [inline src/live.asp.net/Startup.cs]: Sort the `using`

File: src/live.asp.net/Startup.cs [hunk 3]
```diff
@@ -101,7 +102,19 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
...
-            app.UseStaticFiles();
+            app.UseStaticFiles(
+				new StaticFileOptions()
+				{
+					OnPrepareResponse = (context) =>
+					{
+						var headers = context.Context.Response.GetTypedHeaders();
+						headers.CacheControl = new CacheControlHeaderValue()
+						{
+							MaxAge = TimeSpan.FromDays(365)
```
Comment (hishamco) [inline src/live.asp.net/Startup.cs]: I think we can use:
```csharp
var now = DateTime.Now;
MaxAge = now.AddYears(1) - now;
```
instead

File: src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved. 
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using Microsoft.Extensions.FileProviders;
```
Comment (hishamco) [inline src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs]: Sorting!!

File: src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.AspNetCore.Html;
+using Microsoft.Extensions.Options;
+using Microsoft.AspNetCore.Mvc.TagHelpers.Internal;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace live.asp.net.TagHelpers
+{
+	[HtmlTargetElement("link", Attributes = "[rel='stylesheet'],fromCDN")]
+	[HtmlTargetElement("image", Attributes = "fromCDN")]
```
Comment (hishamco) [inline src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs]: IMHO `useCDN` may be better

File: src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.AspNetCore.Html;
+using Microsoft.Extensions.Options;
+using Microsoft.AspNetCore.Mvc.TagHelpers.Internal;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace live.asp.net.TagHelpers
+{
+	[HtmlTargetElement("link", Attributes = "[rel='stylesheet'],fromCDN")]
```
Comment (hishamco) [inline src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs]: Add space after the comma

File: src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	[HtmlTargetElement("image", Attributes = "fromCDN")]
+    public class CDNSourcingTagHelper : TagHelper
+    {
+		private readonly AppSettings _appSettings;
+		private const string FROMCDNAttributeName = "fromCDN";
+
+		public CDNSourcingTagHelper(IOptions<AppSettings> appSettings)
+        {
```
Comment (hishamco) [inline src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs]: Please format the file properly

File: src/live.asp.net/TagHelpers/CacheBustingTagHelper.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	[HtmlTargetElement("*", Attributes = "append-version")]
+    public class CacheBustingTagHelper : TagHelper
+    {
+		private const string AppendVersionAttributeName = "append-version";
+		private FileVersionProvider _fileVersionProvider;
+		protected IHostingEnvironment HostingEnvironment { get; }
+		protected IMemoryCache Cache { get; }
+
+		public CacheBustingTagHelper(IMemoryCache cache, IHostingEnvironment env, IOptions<AppSettings> appSettings)
+        {
```
Comment (hishamco) [inline src/live.asp.net/TagHelpers/CacheBustingTagHelper.cs]: Check the format again ..

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/live.asp.net/AppSettings.cs (hunk 1 above)
Comment (ganniaiuei) [reply src/live.asp.net/AppSettings.cs]: maybe some different settings in the editor, I didnt' see any space in my editor. Should I check something specific? Maybe tab dimension in visualstudio?

File: src/live.asp.net/Startup.cs (hunk 3 above)
Comment (ganniaiuei) [reply src/live.asp.net/Startup.cs]: ok

File: src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs (hunk 5 above)
Comment (ganniaiuei) [reply src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs]: ok

File: src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs (hunk 7 above)
Comment (ganniaiuei) [reply src/live.asp.net/TagHelpers/CDNSourcingTagHelper.cs]: also, in my editor the file is properly formatted.

File: src/live.asp.net/AppSettings.cs (hunk 1 above)
Comment (hishamco) [reply src/live.asp.net/AppSettings.cs]: If you look closely here in the last line you will recognize there's one space missing before `public`, perhaps the Tab Size in VS is different

**Comment by hishamco** [comment]
I think we have spacing issue here 😞

**Comment by alexwiese** [comment]
Looks like a mixture of spaces and tabs to me.  Please reformat using spaces everywhere.

**Comment by ganniaiuei** [comment]
Ok, I've applied a formatting with spaces and the tabs are gone.

**Pull request closed without merging**